File: SampleApps/SampleAppObjC/viewControllers/NoStoryboardViewController.m
//
//  NoStoryboardViewController.m
//  SampleAppObjC
//
//  Copyright © 2017-2020 BEAR SAS. All rights reserved.
//

#import "NoStoryboardViewController.h"
#import "Cache.h"

@interface NoStoryboardViewController ()

@end

@implementation NoStoryboardViewController

-(void)viewDidLoad {
    [super viewDidLoad];
    
    self.scanlineColor = [UIColor greenColor];
    self.recognitionTimeout = 8;
    
    UIButton* close = [UIButton buttonWithType:UIButtonTypeCustom];
    [close setTitleColor:[UIColor systemBlueColor] forState:UIControlStateNormal];
    [close setTitle:@"dismiss" forState:UIControlStateNormal];
    [close sizeToFit];
    [close addTarget:self action:@selector(dismiss) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:close];
    
    close.translatesAutoresizingMaskIntoConstraints = NO;
    
    UILayoutGuide *safeArea = self.view.safeAreaLayoutGuide;
    CGFloat buttonOffset = 20;
    
    [NSLayoutConstraint activateConstraints:@[[close.topAnchor constraintEqualToAnchor:safeArea.topAnchor constant:buttonOffset],
                                              [close.leadingAnchor constraintEqualToAnchor:safeArea.leadingAnchor constant:buttonOffset]]];
}

-(void)dismiss {
    [self dismissViewControllerAnimated:YES completion:nil];
}

-(void)scanTriggered {
    [self.handler startScanning];
}

// MARK: ARDelegate

- (void)recognizedMarkerWithId:(NSInteger)markerId assetIds:(NSArray<NSNumber *> * _Nonnull)assetIds {
    NSLog(@"recognizedMarkerWithId = %ld, assetIds = %@", (long)markerId, assetIds);
    [Cache cacheRecognizedMarkerId:markerId];
}

- (void)recognitionTimeoutReached {
    NSLog(@"recognitionTimeoutReached");
}

- (void)assetClickedWith:(NSInteger)assetId {
    NSLog(@"assetClickedWith assetId - %li", (long)assetId);
}

- (void)viewStateChanged:(enum ARViewState)state {
    NSLog(@"viewStateChanged, state - %li", (long)state);
}

- (void)didFailWithError:(BearError * _Nonnull)error {
    NSLog(@"didFailWithError %@", error);
    UIAlertController* alertVC = [UIAlertController alertControllerWithTitle:error.shortDescription
                                                                     message:error.reason preferredStyle:UIAlertControllerStyleAlert];
    [alertVC addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:alertVC animated:YES completion:nil];
}

@end

File: SampleApps/SampleApp/viewControllers/NoStoryboardViewController.swift
//
//  NoStoryboardViewController.swift
//  SampleApp
//
//  Copyright © 2017-2020 BEAR SAS. All rights reserved.
//

import Foundation
import BearSDK

final class NoStoryboardViewController: ARViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        scanlineColor = .green
        recognitionTimeout = 8
        
        let btn = UIButton(type: .custom)
        btn.setTitle("dismiss", for: .normal)
        btn.setTitleColor(.systemBlue, for: .normal)
        btn.sizeToFit()
        btn.addTarget(self, action: #selector(dismissTapped), for: .touchUpInside)
        view.addSubview(btn)
        
        btn.translatesAutoresizingMaskIntoConstraints = false
        
        let safeArea = view.safeAreaLayoutGuide
        let buttonOffset: CGFloat = 20
        
        NSLayoutConstraint.activate([
            btn.topAnchor.constraint(equalTo: safeArea.topAnchor, constant: buttonOffset),
            btn.leadingAnchor.constraint(equalTo: safeArea.leadingAnchor, constant: buttonOffset)
        ])
    }
    
    @objc func dismissTapped() {
        if let nav = self.navigationController {
            nav.popViewController(animated: true)
        } else {
            self.dismiss(animated: true, completion: nil)
        }
    }
}

extension NoStoryboardViewController: ARDelegate {
    
    func recognizedMarker(withId markerId: Int, assetIds: [Int]) {
        print("called \(#function)")
        cacheRecognized(markerId: markerId)
    }
    
    func recognitionTimeoutReached() {
        print("called \(#function)")
    }

    func assetClicked(with assetId: Int) {
        print("called \(#function)")
    }
    
    func viewStateChanged(_ state: ARViewState) {
        print("called \(#function), state - \(state.rawValue)")
    }
    
    func didFail(withError error: BearError) {
        print("called \(#function)")
    }
}
